oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs

```diff
@@ -0,0 +1,645 @@
+use std::collections::hash_map::RandomState;
+use std::collections::{HashMap, HashSet};
+use std::sync::Mutex;
+
+use parsec_api_types::{BlockID, ChunkID, EntryID};
+use parsec_client_types::LocalManifest;
+
+use super::local_database::{SqliteConn, SQLITE_MAX_VARIABLE_NUMBER};
+use crate::error::{FSError, FSResult};
+use crate::schema::{chunks, prevent_sync_pattern, realm_checkpoint, vlobs};
+
+const EMPTY_PATTERN: &str = r"^\b$"; // Do not match anything (https://stackoverflow.com/a/2302992/2846140)
+
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+pub enum ChunkOrBlockID {
+    ChunkID(ChunkID),
+    BlockID(BlockID),
+}
+
+impl std::convert::AsRef<[u8]> for ChunkOrBlockID {
+    fn as_ref(&self) -> &[u8] {
+        match self {
+            Self::ChunkID(id) => (**id).as_ref(),
+            Self::BlockID(id) => (**id).as_ref(),
+        }
+    }
+}
+
+#[derive(Insertable)]
+#[diesel(table_name = vlobs)]
+struct NewVlob<'a> {
+    vlob_id: &'a [u8],
+    base_version: i32,
+    remote_version: i32,
+    need_sync: bool,
+    blob: &'a [u8],
+}
+
+#[derive(Insertable, AsChangeset)]
+#[diesel(table_name = realm_checkpoint)]
+struct NewRealmCheckpoint {
+    _id: i32,
+    checkpoint: i32,
+}
+
+#[derive(Insertable)]
+#[diesel(table_name = prevent_sync_pattern)]
+struct NewPreventSyncPattern<'a> {
+    _id: i32,
+    pattern: &'a str,
+    fully_applied: bool,
+}
+
+pub struct ManifestStorage {
+    local_symkey: SecretKey,
+    conn: Mutex<SqliteConn>,
+    pub realm_id: EntryID,
+    /// This cache contains all the manifests that have been set or accessed
+    /// since the last call to `clear_memory_cache`
+    pub(crate) cache: Mutex<HashMap<EntryID, LocalManifest>>,
+    /// This dictionnary keeps track of all the entry ids of the manifests
+    /// that have been added to the cache but still needs to be written to
+    /// the conn. The corresponding value is a set with the ids of all
+    /// the chunks that needs to be removed from the conn after the
+    /// manifest is written. Note: this set might be empty but the manifest
+    /// still requires to be flushed.
+    cache_ahead_of_localdb: Mutex<HashMap<EntryID, HashSet<ChunkOrBlockID>>>,
+}
+
+impl Drop for ManifestStorage {
+    fn drop(&mut self) {
+        self.flush_cache_ahead_of_persistance()
+            .expect("Cannot flush cache when ManifestStorage dropped");
+    }
+}
+
+impl ManifestStorage {
+    pub fn new(
+        local_symkey: SecretKey,
+        conn: Mutex<SqliteConn>,
+        realm_id: EntryID,
+    ) -> FSResult<Self> {
+        let instance = Self {
+            local_symkey,
+            conn,
+            realm_id,
+            cache: Mutex::new(HashMap::new()),
+            cache_ahead_of_localdb: Mutex::new(HashMap::new()),
+        };
+        instance.create_db()?;
+        Ok(instance)
+    }
+
+    /// Database initialization
+    pub fn create_db(&self) -> FSResult<()> {
+        let conn = &mut *self.conn.lock().expect("Mutex is poisoned");
```

Comment:
how transactions are handled here ? (in the python code all those sql statements are executed in a single transaction)

Reply:
I decided to run separately in order to handle which table's creation fail since it's called once.
There is batch_execute which perform one transaction